### PR TITLE
QSP angles

### PR DESCRIPTION
**Pull request by renezander90**
## Description

Advances the JAX-based angle computation pipeline for Quantum Signal Processing (QSP) and its variants — GQSP, QSP, and QSVT. The approach is grounded in the equivalence between GQSP and the Non-Linear Fourier Transform (NLFT), using spectral factorization (cepstral method) and a JAX-JIT-compiled layer-stripping algorithm to efficiently compute phase angles from target polynomials. A companion verification module is included for classical simulation and testing of generated angles.

> Note: `xqsp_angles()` and `yqsp_angles()` are present in the module but currently raise `NotImplementedError` — the NLFT-to-X/Y-basis conversion is not yet implemented.

## Related Issues

Closes #
Related to #

## Type of Change

- [x] Feature (new functionality)
- [ ] Change Request (modification of existing functionality)
- [ ] Bug Fix
- [ ] Refactoring (no behavior change)
- [ ] Performance improvement
- [x] Documentation
- [ ] CI / Build

## Breaking Change?
- [ ] Yes
- [x] No

## What was changed?

- **gqsp_angles.py** — core angle computation module:
  - `poly_to_nlft_sequence()`: maps a target polynomial to its NLFT sequence via spectral factorization (`_complementary_polynomial`) and inverse NLFT (`_inlft`, layer-stripping)
  - `gqsp_angles()`: computes `(theta, phi, lambda), alpha` angles for GQSP (Motlagh & Wiebe, PRX Quantum 5, 020368; Laneve, arXiv:2503.03026)
  - `xqsp_angles()` / `yqsp_angles()`: stubs raising `NotImplementedError` (X/Y-restricted angle variants, not yet implemented)
  - `qsp_angles()`: standard QSP phase angles with configurable parity (`even`/`odd`) and signal basis (`X`/`Z`), derived from GQSP via `laurent_to_analytic_coeffs()`
  - `qsvt_angles()`: QSVT phase angles converted from QSP-X angles
  - `laurent_to_analytic_coeffs()`: maps Laurent QSP polynomials to the analytic GQSP framework (Laneve 2025, Lemma 2)
- **gqsp_angles_verification.py** — classical verification utilities:
  - `evaluate_gqsp_polynomial()` / `assert_gqsp_angles_match_target()`
  - `evaluate_nlft_sequence()` / `assert_nlft_sequence_match_target()`
  - `evaluate_qsp_polynomial()` / `assert_qsp_angles_match_target()`
  - `evaluate_qsvt_polynomial()` / `assert_qsvt_angles_match_target()`
  - `plot_reconstruction_vs_target()` for visual diagnostics
- **__init__.py** — exports `gqsp_angles` module symbols
- **test_qsp_angles.py** — parametrized tests for all four angle solvers
- **GQSP.rst** and generated RST stubs for `qsp_angles` and `qsvt_angles` — API reference documentation

## How was it tested?

| Test-ID | Status |
|---------|--------|
| `test_gqsp_polynomial` | ✅ |
| `test_nlft_sequence` | ✅ |
| `test_qsp_angles` | ✅ |
| `test_qsvt_angles` | ✅ |
| `test_xqsp_angles_unimplemented` | ✅ |
| `test_yqsp_angles_unimplemented` | ✅ |

All tests are parametrized across polynomials of varying degree (up to degree 16), including edge cases with sparse coefficients and high-degree monomials. Verification uses direct classical matrix simulation of the QSP/GQSP unitary sequence with `atol=1e-5`.

## Checklist
- [x] My code follows the project's coding standards
- [x] I have performed a self-review
- [x] I have added/updated tests (referencing issue Test-IDs)
- [x] All tests pass locally and in CI
- [x] I have updated the documentation
- [ ] I have added a changelog entry
- [ ] Breaking changes are documented with migration path

## Reviewer Notes

- `xqsp_angles()` and `yqsp_angles()` unconditionally raise `NotImplementedError`. They are included as stubs for future implementation; the NLFT → X/Y-basis conversion logic exists in private helpers but is currently disabled.
- The layer-stripping loop in `_inlft` is JAX-JIT-compiled via `jax.lax.while_loop` for performance on large-degree polynomials.

### CODE REVIEW

**Comment by MatP1337** [comment]
lfg, awesome!

File: src/qrisp/algorithms/gqsp/gqsp_angles.py [hunk 1]
```diff
@@ -212,7 +212,7 @@ def _gqsp_angles_from_nlft_sequence(F: Array) -> Tuple[Array, Array, Array]:
     Parameters
     ----------
     F : ArrayLike
-        1-D array containing the polynomial coefficients, ordered from lowest order term to highest.
+        1-D array containing the sequence, ordered from lowest order term to highest.
```
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles.py]: What kind of sequence are we dealing here? Should be more descriptive imho

File: src/qrisp/algorithms/gqsp/gqsp_angles.py [hunk 2]
```diff
@@ -300,10 +323,233 @@ def gqsp_angles(p: "ArrayLike") -> Tuple[Tuple[Array, Array, Array], Array]:
...
+
+    """
+    F, alpha = poly_to_nlft_sequence(p)
     theta, phi, lambda_ = _gqsp_angles_from_nlft_sequence(F)
 
     return (theta, phi, lambda_), alpha
+
+
+# https://arxiv.org/pdf/2503.03026 
+# Not verified to be correct.
```
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles.py:382]: Should be emphasized more. It's nice to have the entire transformation ZOO included, but based on discussion with @renezander90, perhaps it's better to mark this as internal function until the function is verified to work as intended.

For the end user, this doesn't change anything since they will just call the soon-to-be adapted QSVT function, obtaining angles from gqsp by setting alpha and lambda to 0 with the appropriate changes to the phis.

File: src/qrisp/algorithms/gqsp/gqsp_angles.py [hunk 3]
```diff
@@ -300,10 +323,233 @@ def gqsp_angles(p: "ArrayLike") -> Tuple[Tuple[Array, Array, Array], Array]:
...
+def xqsp_angles(p: "ArrayLike") -> Tuple[Array, Array]:
```
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles.py:383]: see comment above (R366)

File: src/qrisp/algorithms/gqsp/gqsp_angles.py [hunk 4]
```diff
@@ -300,10 +323,233 @@ def gqsp_angles(p: "ArrayLike") -> Tuple[Tuple[Array, Array, Array], Array]:
...
+
+    """
+    F, alpha = poly_to_nlft_sequence(p)
+    phi = _xqsp_angles_from_nlft_sequence(F)
+
+    return phi, alpha
+
+
+# https://arxiv.org/pdf/2503.03026 
+# Not verified to be correct.
```
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles.py:419]: same thing as above (R366&R367)

File: src/qrisp/algorithms/gqsp/gqsp_angles.py [hunk 5]
```diff
@@ -300,10 +323,233 @@ def gqsp_angles(p: "ArrayLike") -> Tuple[Tuple[Array, Array, Array], Array]:
...
+def yqsp_angles(p: "ArrayLike") -> Tuple[Array, Array]:
```
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles.py:420]: same thing as above (R366&R367)

File: src/qrisp/algorithms/gqsp/gqsp_angles_verification.py
Comment (MatP1337) [inline src/qrisp/algorithms/gqsp/gqsp_angles_verification.py:1]: nice and thorough, awesome!

File: tests/algorithms_tests/qsp_tests/test_qsp_angles.py
Comment (MatP1337) [inline tests/algorithms_tests/qsp_tests/test_qsp_angles.py:1]: Looks good 🤙🏻

**Review by MatP1337 — CHANGES_REQUESTED** [review]
Looks good, nice to have the QSP and QSVT angles available for the users without the utilization of other libraries.

Makes it simple to implement the QSVT with the angles hidden from the user.

ONLY REMARK: regarding the transformations that are ``# not verified`` - they should be marked as internal WIP functions. In the end, there is a modest amount of redundancy since one can obtain the same angles also via gqsp -> qsp -> qsvt.

Imho it's good to have the full picture implemented - we can continue and/or verify the other transformations post v0.9. Thoughts/opinions @renezander90 @positr0nium 

all in all, great addition, branching out from this branch for utilizing these angles for QSVT asap.

Kudos @renezander90 🙌🏻

File: src/qrisp/algorithms/gqsp/gqsp_angles.py (hunk 2 above)
Comment (renezander90) [reply src/qrisp/algorithms/gqsp/gqsp_angles.py:382]: Now raises NotImplementedError

File: src/qrisp/algorithms/gqsp/gqsp_angles.py (hunk 3 above)
Comment (renezander90) [reply src/qrisp/algorithms/gqsp/gqsp_angles.py:383]: Now raises NotImplementedError

File: src/qrisp/algorithms/gqsp/gqsp_angles.py (hunk 4 above)
Comment (renezander90) [reply src/qrisp/algorithms/gqsp/gqsp_angles.py:419]: Now raises NotImplementedError

File: src/qrisp/algorithms/gqsp/gqsp_angles.py (hunk 5 above)
Comment (renezander90) [reply src/qrisp/algorithms/gqsp/gqsp_angles.py:420]: Now raises NotImplementedError

File: src/qrisp/algorithms/gqsp/gqsp_angles.py (hunk 1 above)
Comment (renezander90) [reply src/qrisp/algorithms/gqsp/gqsp_angles.py]: non-linear Fourier sequence

**Review by MatP1337 — APPROVED** [review]
Looking good! Thanks for addressing the review comments. I think the PR is now ready to be merged into main. Awesome work @renezander90 (as per usual)!!

**Comment by MatP1337** [comment]
yesyesyes
